### DISCUSSION TITLE
Rich type information for runtime metadata using `scale-info`

### DISCUSSION THREAD

**Discussion by ascjones** [paritytech/substrate#8370] [General]
Opening a discussion for tracking and feedback on the proposed integration of `scale-info` into substrate.

Currently the substrate metadata doesn't contain full type information, so extra mappings to type definitions are required in client implementations e.g. polkadot.js in order to to provide compatible encoding/decoding of types when interacting with substrate.

The goal is to be able to have enough information embedded in the metadata in order to generate compatible types in any language. This will greatly reduce the maintenance burden for client implementations.

The plan is to use [`scale-info`](https://github.com/paritytech/scale-info) to derive type information for all types that are exposed in a substrate runtime via calls, events. storage, constants etc.

## Current status

- Proof of concept branch on substrate: https://github.com/paritytech/substrate/compare/aj-metadata-vnext. I will post an example of the metadata produced by this soon to get some feedback.
- New metadata generation depends on FRAME V2 macros, so depends on all pallets being converted: https://github.com/paritytech/substrate/issues/7882 
- `scale-info` itself still needs a feature or two before it is ready: https://github.com/paritytech/scale-info/issues/36

Once all FRAME V2 macros are converted and final `scale-info` features are added I will open a substrate PR based on my PoC branch.

### Should the metadata still be stored in the runtime itself?

Currently the metadata is stored in the runtime Wasm blob, however this could become unwieldy with the extra type information. Still need to measure the size difference, it might not be that bad but something to keep in mind. One idea just store a hash of the metadata in the runtime itself and the metadata itself somewhere else (where?).

### Backwards compatibility

It would be possible to provide a conversion from the new style metadata (with full type info) to the deprecated style (without full type info) to ease the transition in the client ecosystem. Do we want to do this?

## Related issues:

https://github.com/paritytech/substrate/issues/917
https://github.com/paritytech/substrate/issues/8158
https://github.com/paritytech/substrate/issues/8083

/cc @Robbepop @dvdplm @thiolliere @tomusdrw @kianenigma @bkchr @jacogr @xlc

Please tag anyone else who would be interested in this discussion.

**Comment by gui1117** [comment]
> New metadata generation depends on FRAME V2 macros, so depends on all pallets being converted: #7882

Isn't it possible to have an empty implementation of new metadata for old macro? so that pallet with old macro can still be used but only provide old metadata?

**Reply by ascjones** [reply]
Yes it is possible, I originally expected that we would have everything converted sooner which is why I didn't consider it originally.

**Reply by apopiak** [reply]
Even if we manage to convert in due time, it will take time for the ecosystem to do the same.

**Reply by ascjones** [reply]
In the end I converted the legacy macros to produce the new metadata, so now we have metadata even for the non converted pallets.

**Comment by tomusdrw** [comment]
> One idea just store a hash of the metadata in the runtime itself and the metadata itself somewhere else (where?).

That sounds reasonable. I think we could use some existing Substrate's Off-Chain capabilities to store the metadata. In any case I think it's essential to split runtime code and the metadata into two separate blobs.
1. The runtime upgrade transaction contains `code` and `metadata` blobs as separate parameters.
1.1. Metadata is stored in the state (the code might optionally contain the hash to verify correctness).
1.2. Only the hash of the metadata is stored in the state and the content itself is written to Off-Chain DB via indexing API.
1. Runtime upgrade transaciton contains only `code` and the hash of the metadata.
2.1. Correct metadata can be then provided in a separate transaction (could even be automatically triggered by an Off-Chain Worker) and similarly:
2.2. Either stored on-chain.
2.3. or written to the Off-Chain DB via indexing API.


Having the metadata in the Off-Chain DB makes it easy for all consumers (like UI) to query it using built-in RPC method. Usage of Indexing API ensures correctness of the data.

**Reply by bkchr** [reply]
And how would the `metadata` generated in this case? 

How would this work for dev runtimes? 

And I would like to not depend on offchain stuff for the metadata, while it sounds nice. Offchain workers are not required by FRAME.

**Reply by ascjones** [reply]
> And how would the metadata generated in this case?

I'm not 100% familiar with the runtime build process, but would it be possible to something like we do in `ink!` where we pre-generate the metadata by invoking the (`std` only) `generate_metadata()` function and encoding it. Note that in our case we use `cargo-contract` to generate a temporary cargo project to invoke this.

**Reply by tomusdrw** [reply]
> And how would the metadata generated in this case?

Same as we generate runtime code blob. The build will create two artifacts `code.wasm` and `metadata.blob`, which you either submit both in a single transaction or separately.

> How would this work for dev runtimes?

Genesis could have the `metadata` on-chain storage entry populated via chain spec file with full metadata or just a hash (depending on the chosen path). In case of a hash we could also special-case the node binary to submit a transaction in the first block (if the runtime version matches) with full metadata content.
From my PoV the node & runtime should be fully functional with or without metadata.

> And I would like to not depend on offchain stuff for the metadata, while it sounds nice. Offchain workers are not required by FRAME.

Metadata is not really required by FRAME either? Note that we have two options here we can store on-chain or off-chain, I think it's just crucial to separate metadata from runtime code blob for easier consumption.
Using Off-Chain DB is IMHO super nice, cause we don't pollute on-chain storage, while at the same time make it really easy for UI/client libs to retrieve the metadata without requiring them to query some other distributed-storage solutions (like IPFS).

**Reply by tomaka** [reply]
How would a light client know the metadata if it's only stored in some off-chain DB?
And this also means that we lose older metadatas?

The regular on-chain storage solution looks much more sane to me.

**Reply by apopiak** [reply]
+1 for keeping metadata and runtime separately and +1 for having both in storage

**Reply by tomusdrw** [reply]
> How would a light client know the metadata if it's only stored in some off-chain DB?

Since the data you receive should be on-chain verifiable anyway (i.e. compare with hash) I can imagine this being either:
1. Not our problem (it can be solved by being hardcoded, IPFS, centralised, querying some off-chain enabled RPC nodes)
2. Being requested from other nodes via some networking req-res protocol. Quite similar to how you'd receive an on-chain data anyway. (obviously a bit more work to implement such protocol).

> And this also means that we lose older metadatas?

In case of using Off-Chain DB, not at all. Each metadata can have it's own entry (hash being the key for instance). If we store on-chain, I don't think we can afford that though.

> The regular on-chain storage solution looks much more sane to me.

Sure that's the most straight-forward, although it means that: (1) the metadata needs to be delivered on-chain (large transaction) (2) and it has to be stored in the trie (we need to avoid shared prefix, but obviously doable if we don't store historic data).

**Reply by tomaka** [reply]
Why not put the runtime code itself in the offchain storage, then?

**Reply by tomusdrw** [reply]
Could work, yeah. I guess the main difference is that runtime code is essential for full nodes (network security), while metadata is only required for user interactions, hence I'm more on the safe side with runtime code, but not metadata.

**Reply by ascjones** [reply]
Reviving this thread, given results of polkadot runtime binary size increase (2M -> 2.4M minimum) https://github.com/paritytech/substrate/discussions/8370#discussioncomment-909365.

We need to consider whether we need to do the separation, if so where to store it, and whether to do it as part of https://github.com/paritytech/substrate/pull/8615, or as a follow up.

**Reply by apopiak** [reply]
@bkchr Seems like this would create further issues with oversized runtimes?

**Comment by tomusdrw** [comment]
> It would be possible to provide a conversion from the new style metadata (with full type info) to the deprecated style (without full type info) to ease the transition in the client ecosystem. Do we want to do this?

Can you elaborate how is this supposed to look like? If it's just about adding new version, then I think it's acceptable - i.e. metadata was designed to be upgraded. If the compatibility is trivial though, we might implement it to make life of client developers easier, especially to not break support for hardware wallets.

**Reply by ascjones** [reply]
Haven't put a lot of thought into that, which is why it is an open question. I assume there would be a routine provided in `frame-metadata` that would be `From<RuntimeMetadataV13> for RuntimeMetadataV12`. Then we could possibly provide that as either an additional "deprecated" rpc endpoint for convenience (which has it's own problems), or just provide it as a "tool" in `frame-metadata` that the clients will have to run locally to do the conversion.

**Reply by dvdplm** [reply]
Some tooling will be interested in old versions of metadata (archive tools, chain analytics, …), which is an argument for storing it off-chain (or even outside the node)?

DQ: Do we really care about defining what "correct" metadata is? Are there any attacks possible stemming from using "incorrect" metadata? To me it sounds like it's a "if it works it works"-situation where it doesn't really matter what version of the metadata you're using as long as it can decode/encode the SCALE encoded data. If I were to send an extrinsic encoded using the wrong metadata, what would the consequences be?

The context for my question is tools like Parity Signer that needs a way to update its metadata by scanning QR codes (or some other non-networked solution) but that in reality only needs a subset of the metadata, enough to do its limited job of signing. Or Ledger devices that are severely space-constrained and will want to use as small a payload as possible.

It would be good to have a lot of freedom in how different use-cases deploy metadata.

**Reply by gui1117** [reply]
I thought parity signer needs metadata in order to show details about the transaction the user is signing.
In this context the detail seems important to be correct. if a transaction say argument 1 is `transfer_to` and argument 2 is `transfer_from` then it is important to ensure the metadata gives correct information.

**Reply by tomusdrw** [reply]
> Do we really care about defining what "correct" metadata is? 

I guess it's the matter of "the source of truth". It's not strictly required for the UIs/client libs to have the correct metadata information (see what happens if polkadot.js/apps doesn't have all/correct custom types), but they might simply randomly break or display incorrect data. In an extreme case they might create a transaction that is mis-interpreted by the chain (For instance sending `6` tokens to `12345` address might end up being a transfer for `56` tokens to `1234` address), so IMHO for some apps it's super important to be able to verify metadata correctness.
Regarding where to get the metadata from, I also think it should be as simple as possible and should be coming from the node you are connected to - it doesn't necessarily have to be on-chain, hence I was proposing off-chain storage with on-chain verifiability. Given a hash, the metadata can also be retrieved by other means (for instance IPFS).
Using PJS/apps with custom types right now is really annoying, it's not clear where to get the custom types from and the UI is not able to auto-apply them, having no single source of metadata information will imho end up being the same flaky user experience.

**Reply by dvdplm** [reply]
>  if a transaction say argument 1 is `transfer_to` and argument 2 is `transfer_from` then it is important to ensure the metadata gives correct information.

This is true today as well though right? If I use v13 transaction version to encode my transfer and submit it to a node running v14 where the params switched place, it'll fail (as it should). If I use "homecooked" (or maliciously crafted) type data today all bets are off and if the extrinsic can be decoded by the node it will be executed.

I am not sure how big of a problem this actually is though.

Maybe it's worth splitting the problem into two:
1. metadata distribution – make using metadata less difficult and less of a manual operation
2. attesting the source of truth – ensuring that the metadata you have is what it claims to be

Can they be tackled separately?

**Reply by tomusdrw** [reply]
> Can they be tackled separately?

Yes, sure, but (2) should be simple, hence the suggestions that the on-chain component (runtime code, storage or both) should only contain the hash (i.e. be able to verify the correctness of the metadata) and the content itself could be distributed off-chain in some form that is more or less closely coupled with the runtime itself.

**Reply by athei** [reply]
Maybe we can even modularize the metadata in order to allow for a smaller payload for constrained signer devices. Signers would only be interested in metadata regarding extrinsincs but not the new type data that also covers RPC signatures.

**Reply by bkchr** [reply]
RPC signatures? :D They are clearly not part of the metadata :D

**Reply by jacogr** [reply]
No they are not, but probably should have the same structure. They have types as well, and the intent has always been to make them self-descriptive, hence the version in `rpc_methods`, but we are not there.

That _may_ actually be a small, less-impact place to start. (Certainly less annotations all over the codebase for RPC-only and can probably keep the existing info alongside providing types on the same RPC there)

**Reply by athei** [reply]
> RPC signatures? :D They are clearly not part of the metadata :D

I thought this is meant by "adding rich type information to substrate".

**Reply by bkchr** [reply]
No :) These are the same type information as extracted from ink contracts.

**Comment by xlc** [comment]
Just one thing about timing. If we can get this out before Kusama parachain launch, this will mean all parachain projects will not need to start with the legacy metadata system. This definitely will make a lot of people's life better (including myself) so that we don't need to have single line of code dealing with the legacy system and all the migration.

So I am keen to do whatever possible to get this out.

Currently parachain custom types are manually generated, hard to validate, centralized discovery & distribution, and no cryptographic authentication anyway, so I don't think we can do anything worse than that.

My answer to the questions:

**Should the metadata still be stored in the runtime itself?**

No. The types are not stored in current runtime anyway so that's not a regression. Metadata are huge and we already have enough trouble to keep our wasm size small so it play nicely with parachain production code. I think it is a mistake to have metadata include all the comments already.

**Backwards compatibility**

Bumping metadata version will causing similar breakage everywhere. We already have multiple runtime breaks balances display (all the ref count changes). I know wallet & exchanges won't like this, but this is the solution to reduce breaking changes so I will 100% say just break it.

**Reply by olanod** [reply]
+1 for keeping WASM size tiny, I like the idea of only keeping a hash that points to e.g. an IPFS CID. Would the [storage chains](https://github.com/paritytech/substrate/issues/7962) stuff be helpful there if nodes become "IPFS ready"? could metadata be something kept only in the node client side only to be served via RPC or to other IPFS nodes and the runtime not care about it?

**Comment by jacogr** [comment]
This is not the first metadata bump and won't be the last. For instance any API out there (Go, Ruby, Python, JS, ...) already needs to support at least metadata versions 9, 10, 11 and 12 (current) to be able to parse all historic data from Kusama. 

The sad thing is that because of the historic data, v13 won't remove the old custom types from any of the APIs (historic support), however at least going forward we won't need to add more. Which is an absolutely brilliant place to be. (And all new chain, as @xlc said, won't have this burden at all)

**Comment by emielsebastiaan** [comment]
Polkascan Foundation fully supports Bryan's (@xlc) suggestions.

**Comment by ascjones** [comment]
> New metadata generation depends on FRAME V2 macros, so depends on all pallets being converted: #7882

Update: I have now converted the legacy frame macro to produce the new metadata in https://github.com/paritytech/substrate/compare/aj-metadata-vnext, so it can now produce the scale-info enriched metadata for all pallets.

See here for the resulting metadata for `node-runtime`: https://gist.github.com/ascjones/d0d60da26f4336a9379547a2ea1d5655. Note that only `calls` and `events` are populated, still need to do storage, constants and errors.

**Reply by ascjones** [reply]
Example of complete `SCALE` compatible rust types generated from the above metadata: https://github.com/ascjones/chameleon/blob/master/examples/codegen/src/substrate_node_runtime.rs

**Comment by olanod** [comment]
This is going to be big for us as well, we are building a plugin based "decentralizable" API for our app and its blockchain/substrate plugin is based on [`sube`](https://github.com/valibre-org/sube) that I created as mini alternative to `subxt` pretty much based on the assumption that v13 metadata could come soonish. If there are blockers like migrating the pallets we could help a bit there and allocate some time to speed things up. The killer feature for _sube_ will be converting `JSON` or plain text to `SCALE` with the only requirement being a node URL to fetch metadata or directly passing the SCALE blob of the metadata which by the way would be nice if it can be [lighter weight](https://github.com/paritytech/frame-metadata/issues/6)

**Comment by ascjones** [comment]
Early draft PR: https://github.com/paritytech/substrate/pull/8615

**Comment by gregdhill** [comment]
Any updates @ascjones? We (@Interlay) are anxious to integrate these changes into our clients before Kusama.

**Reply by ascjones** [reply]
My goal is to have the PR ready to review by the end of this week, just starting on the `polkadot` companion PR now

**Reply by ascjones** [reply]
Yeah okay it took a bit longer than I expected :see_no_evil: https://github.com/paritytech/polkadot/pull/3336

**Comment by ascjones** [comment]
First run of generating new metadata for polkadot: https://gist.github.com/ascjones/726caaf29eff76ca72f142f419e9c6a9

**Reply by tomaka** [reply]
When it says for example `"ty:" { "Plain": 4 }`, does that mean that it refers to the 4th type in the list (here `u32`)?

**Reply by ascjones** [reply]
Exactly it's the index into the `types` array

**Comment by ascjones** [comment]
Regarding Wasm binary size, here are some numbers for the polkadot runtime:

- 2.0M (Existing)
- 2.7M (New)
- 2.4M (New without docs)

So a minimum of a 20% increase in the size of the runtime which includes this type metadata.

**Reply by xlc** [reply]
For context of parachains, what's the compressed size?

**Reply by emielsebastiaan** [reply]
I suppose I do not have to go on the record that this feature is at the very top of our wish list.

Documentation is not functional in the metadata and the type registry/decomposition is functional in the metadata.

**Reply by ascjones** [reply]
> For context of parachains, what's the compressed size?

How do I run the compression? I built these with https://github.com/paritytech/polkadot/blob/c69ba7e2d4b4032dab11939129d890bfc88e210e/scripts/build-only-wasm.sh

**Reply by tomaka** [reply]
The `.wasm` is compressed using the `zstd` algorithm. You don't necessarily have to use the same code as Substrate, just compressing manually should give you the same compression level.

**Reply by ascjones** [reply]
I just used the `zstd` CLI to compress those, and the result was:

- 608K (Existing)
- 748K (New)
- 684K (New without docs)

**Reply by ascjones** [reply]
https://github.com/paritytech/substrate/pull/8615#issuecomment-887281108

**Comment by ascjones** [comment]
Marked https://github.com/paritytech/substrate/pull/8615 ready for review, still working on fixing some tests and size benchmarking, but ready for some feedback.
